ams/server/src/main/java/com/netease/arctic/server/optimizing/UnKeyedTableCommit.java

```diff
@@ -105,41 +109,60 @@ protected List<DataFile> moveFile2HiveIfNeed() {
-                  client.getTable(table.id().getDatabase(), table.id().getTableName()));
-              partitionPath = hiveTable.getSd().getLocation();
-            } catch (Exception e) {
-              LOG.error("Get hive table failed", e);
-              throw new RuntimeException(e);
-            }
-          } else {
-            String hiveSubdirectory = table.isKeyedTable() ?
-                HiveTableUtil.newHiveSubdirectory(maxTransactionId) : HiveTableUtil.newHiveSubdirectory();
-
-            Partition partition = HivePartitionUtil.getPartition(hiveClient, table, partitionValues);
-            if (partition == null) {
-              partitionPath = HiveTableUtil.newHiveDataLocation(((SupportHive) table).hiveLocation(),
-                  table.spec(), targetFile.partition(), hiveSubdirectory);
-            } else {
-              partitionPath = partition.getSd().getLocation();
-            }
-          }
-          partitionPathMap.put(taskRuntime.getPartition(), partitionPath);
-        }
+        String partitionPath = partitionPathMap.computeIfAbsent(taskRuntime.getPartition(),
+            key -> getPartitionPath(hiveClient, maxTransactionId, targetFile, partitionSchema));
 
-        DataFile finalDataFile = moveTargetFiles(targetFile, partitionPathMap.get(taskRuntime.getPartition()));
+        DataFile finalDataFile = moveTargetFiles(targetFile, partitionPath);
         newTargetFiles.add(finalDataFile);
       }
     }
     return newTargetFiles;
   }
 
+  private String getPartitionPath(HMSClientPool hiveClient, long maxTransactionId, DataFile targetFile,
+                                  Types.StructType partitionSchema) {
+    // get iceberg partition path
+    String icebergPartitionLocation = getIcebergPartitionLocation(targetFile.partition());
+    if (icebergPartitionLocation != null) {
+      return icebergPartitionLocation;
+    }
+    // get hive partition path
+    if (table.spec().isUnpartitioned()) {
+      try {
+        Table hiveTable = ((SupportHive) table).getHMSClient().run(client ->
+            client.getTable(table.id().getDatabase(), table.id().getTableName()));
+        return hiveTable.getSd().getLocation();
+      } catch (Exception e) {
+        LOG.error("Get hive table failed", e);
+        throw new RuntimeException("Get hive table failed", e);
+      }
+    } else {
+      List<String> partitionValues = HivePartitionUtil.partitionValuesAsList(targetFile.partition(), partitionSchema);
+      String hiveSubdirectory = table.isKeyedTable() ?
+          HiveTableUtil.newHiveSubdirectory(maxTransactionId) : HiveTableUtil.newHiveSubdirectory();
+
+      Partition p = HivePartitionUtil.getPartition(hiveClient, table, partitionValues);
+      if (p == null) {
+        return HiveTableUtil.newHiveDataLocation(((SupportHive) table).hiveLocation(),
+            table.spec(), targetFile.partition(), hiveSubdirectory);
+      } else {
+        return p.getSd().getLocation();
+      }
```

Comment:
Does it seem like there might be a performance issue if HMSClient.getPartition() needs to be called for each DataFile?

Reply:
The `computeIfAbsent` is trying to avoid the performance issue.
```
String partitionPath = partitionPathMap.computeIfAbsent(taskRuntime.getPartition(),
            key -> getPartitionPath(hiveClient, maxTransactionId, targetFile, partitionSchema));
```